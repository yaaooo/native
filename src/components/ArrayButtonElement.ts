import { VNode } from "vue";
import { ArrayButtonComponent } from "../../types";

export const ArrayButtonElement: ArrayButtonComponent = {
  functional: true,
  name: "ArrayButtonElement",
  render(h, { props: { button } }): VNode | VNode[] {
    return h("button", {
      attrs: {
        type: "button",
        title: button.tooltip,
        disabled: button.disabled,
        "data-fs-button": button.type
      },
      on: {
        click: button.trigger as any
      }
    }, button.label);
  }
};
