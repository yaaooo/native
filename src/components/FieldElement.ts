import { VNode } from "vue";
import { Field } from "@/lib/Field";
import { Dict, FieldComponent } from "../../types";

export const FieldElement: FieldComponent = {
  name: "FieldElement",
  functional: true,
  render(h, { data, props, children }): VNode | VNode[] {
    const field = props.field;
    const descriptor = props.field.descriptor;

    if (field.kind === "hidden" || descriptor.definition.kind === "hidden") {
      return children;
    }

    const type = descriptor.attrs.type || field.kind;
    const attrs: Dict = {
      "data-fs-kind": field.kind,
      "data-fs-type": type,
      "data-fs-field": field.name,
      "data-fs-required": field.required,
      "data-fs-horizontal": descriptor.props.horizontal
    };

    const labelElement = h("label", {
      attrs: descriptor.labelAttrs
    }, descriptor.label);

    const fieldElement = h("div", {
      attrs: {
        "data-fs-input": type
      },
      props: { field }
    }, children);

    const nodes = [ fieldElement ];
    const helperNode = h(descriptor.components.get("helper"), data);

    if (helperNode.tag) {
      nodes.push(helperNode);
    }

    Field.renderMessages(h, field, nodes);

    const wrapperElement = nodes.length === 1
      ? nodes
      : h("div", {
        attrs: {
          "data-fs-wrapper": nodes.length
        }
      }, nodes);

    return h("div", { attrs }, [ labelElement, wrapperElement ]);
  }
};
