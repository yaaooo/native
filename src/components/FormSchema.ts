import { VNode } from "vue";
import { CreateElement } from "vue/types/umd";
import { UniqueId as UniqueIdLib } from "@/lib/UniqueId";
import { Objects as ObjectsLib } from "@/lib/Objects";
import { Components as ComponentsLib } from "@/lib/Components";
import { Fieldset as FieldsetLib } from "@/lib/Fieldset";
import { Parser as ParserLib } from "@/parsers/Parser";
import { NativeElements as NativeElementsLib } from "@/lib/NativeElements";
import { NativeComponents as NativeComponentsLib } from "@/lib/NativeComponents";
import { UIDescriptor as UIDescriptorLib } from "@/descriptors/UIDescriptor";

import "@/parsers";
import "@/descriptors";

import {
  ComponentOptions,
  RenderContext,
  WatchOptions,
  ComputedOptions
} from "vue/types/options";

import {
  SubmitEvent,
  ParserOptions,
  UnknowField,
  DescriptorInstance,
  Dict,
  IComponents,
  IParser,
  IUIDescriptor,
  ValidatorFunction
} from "../../types";

import { JsonSchema } from "../../types/jsonschema";

/**
 * FormSchema API
 */
interface FormSchemaVue extends Vue {
  // props
  schema: JsonSchema;
  value?: unknown;
  id: string;
  name?: string;
  bracketedObjectInputName: boolean;
  search: boolean;
  disabled: boolean;
  components: IComponents;
  descriptor: DescriptorInstance | IUIDescriptor;
  validator: ValidatorFunction;

  // data
  key: string;
  ref: string;
  initialModel: unknown;
  ready: boolean;
  parser: IParser<any, UnknowField> | null;

  // computed
  fieldId: string;
  listeners: Record<string, Function | Function[]>;

  // methods
  clone(value: unknown): unknown;
  form(): HTMLFormElement | VNode | undefined;
  emitInputEvent(value: unknown, field: UnknowField): void;
  update(updatedFields: UnknowField[]): void;
}

type Accessors<T, V> = {
  [K in keyof T]: ((this: V) => T[K]) | ComputedOptions<T[K]>
}

type WatchHandler<V extends Vue, T> = (this: V, val: T, oldVal: T) => void;

interface WatchOptionsWithHandler<V extends Vue, T> extends WatchOptions {
  handler: WatchHandler<V, T>;
}

type Props = Record<string, any>;

interface FormSchemaComponent<V extends FormSchemaVue = FormSchemaVue> extends ComponentOptions<V> {
  computed: Accessors<Dict, V>;
  watch?: Record<string, WatchOptionsWithHandler<V, any> | WatchHandler<V, any> | string>;

  render?(this: V, createElement: CreateElement, hack: RenderContext<Props>): VNode;
  renderError?(this: V, createElement: CreateElement, err: Error): VNode;
  staticRenderFns?: ((this: V, createElement: CreateElement) => VNode)[];

  beforeCreate?(this: V): void;
  created?(this: V): void;
  beforeDestroy?(this: V): void;
  destroyed?(this: V): void;
  beforeMount?(this: V): void;
  mounted?(this: V): void;
  beforeUpdate?(this: V): void;
  updated?(this: V): void;
  activated?(this: V): void;
  deactivated?(this: V): void;
  errorCaptured?(err: Error, vm: Vue, info: string): boolean | void;
  serverPrefetch?(this: V): Promise<void>;
}

export const GLOBAL = {
  Elements: Object.freeze(NativeElementsLib)
};

export const Objects = ObjectsLib;
export const UniqueId = UniqueIdLib;
export const Components = ComponentsLib;
export const Fieldset = FieldsetLib;
export const Parser = ParserLib;
export const UIDescriptor = UIDescriptorLib;
export const NativeComponents = NativeComponentsLib;

/**
 * @slot default - Use default slot to insert custom form buttons
 */
const FormSchema: FormSchemaComponent = {
  name: "FormSchema",
  model: {
    prop: "value",
    event: "input"
  },
  props: {
    /**
     * The input JSON Schema object.
     */
    schema: {
      type: Object,
      required: true
    },

    /**
     * Use this directive to create two-way data bindings with the
     * component. It automatically picks the correct way to update the
     * element based on the input type.
     *
     * @type any
     */
    value: {
      type: [ Number, String, Array, Object, Boolean ],
      default: undefined
    },

    /**
     * The id property of the Element interface represents the form's identifier,
     * reflecting the id global attribute.
     *
     * @default Random unique ID
     */
    id: {
      type: String,
      default: UniqueId.get("form")
    },

    /**
     * The name of the form. It must be unique among the forms in a document.
     */
    name: {
      type: String,
      default: undefined
    },

    /**
     * When set to `true` (default), checkbox inputs and nested object inputs
     * will * automatically include brackets at the end of their names
     * (e.g. `name="grouped-checkbox-fields[]"`).
     * Setting this property to `false`, disables this behaviour.
     */
    bracketedObjectInputName: {
      type: Boolean,
      default: true
    },

    /**
     * Use this prop to enable `search` landmark role to identify a section
     * of the page used to search the page, site, or collection of sites.
     */
    search: {
      type: Boolean,
      default: false
    },

    /**
     * Indicates whether the form elements are disabled or not.
     */
    disabled: {
      type: Boolean,
      default: false
    },

    /**
     * Use this prop to overwrite the default Native HTML Elements with
     * custom components.
     *
     * @default GLOBAL.Elements
     */
    components: {
      type: Components,
      default: (): Readonly<NativeComponentsLib> => GLOBAL.Elements
    },

    /**
     * UI Schema Descriptor to use for rendering.
     *
     * @type DescriptorInstance
     * @typeref #descriptor-interface
     * @default {}
     */
    descriptor: {
      type: Object,
      default: (): object => ({})
    },

    /**
     * The validator function to use to validate data before to emit the
     * `input` event.
     *
     * @see [Custom Validation API](#custom-validation-api)
     * @kind function
     * @param {GenericField} field - The field that requests validation
     * @param {string} field.id - The input ID attribute value
     * @param {string} field.name - The input name attribute value
     * @param {any} field.value - The input value for validation
     * @param {JsonSchema} field.schema - The JSON Schema object of the input
     * @param {boolean} field.required - Boolean indicating whether or not the field is mandatory
     * @param {boolean} field.hasChildren - Boolean indicating whether or not the field has children
     * @param {any} field.initialValue - The initial input value
     * @param {Message[]} field.messages - The input value for validation
     * @returns {Promise<boolean>} A promise that return `true` if validation success and `false` otherwise
     */
    validator: {
      type: Function,
      default: null
    }
  },
  data: () => ({
    key: undefined,
    ref: UniqueId.get("formschema"),
    initialModel: undefined,
    ready: false,
    parser: null
  }),
  computed: {
    fieldId(): string {
      return `${this.id}-field`;
    },
    options(): ParserOptions<any, any> {
      return {
        schema: this.schema,
        model: this.initialModel,
        name: this.name,
        id: this.fieldId,
        required: true,
        descriptor: this.descriptor,
        components: this.components,
        bracketedObjectInputName: this.bracketedObjectInputName,
        onChange: this.emitInputEvent,
        validator: this.validator,
        requestRender: this.update
      };
    },
    listeners(): Record<string, Function | Function[]> {
      const on: any = { ...this.$listeners };

      on.reset = on.reset ? [ on.reset ] : [];

      on.reset.unshift(() => {
        if (this.parser) {
          this.parser.field.reset();
        }
      });

      if (on.submit) {
        const onsubmit = on.submit;

        on.submit = (event: SubmitEvent) => {
          if (this.parser) {
            event.field = this.parser.field;
          }

          return onsubmit(event);
        };
      }

      // remove the injected vue's input event
      // to prevent vue errors on the submit event
      delete on.input;

      return on;
    }
  },
  watch: {
    schema: {
      immediate: true,
      handler(): void {
        this.ready = false;
        this.initialModel = Objects.clone(this.value as any);
      }
    },
    options: {
      immediate: true,
      handler(options: ParserOptions<any, any>): void {
        this.parser = Parser.get(options);
      }
    }
  },
  render(createElement): VNode {
    if (this.parser === null || this.ready === false) {
      return null as any; // nothing to render
    }

    const attrs = {
      disabled: this.disabled
    };

    const props = {
      field: this.parser.field
    };

    const element = createElement(props.field.descriptor.component, { attrs, props });
    const nodes = [ element ];

    if (this.$slots.default) {
      nodes.push(...this.$slots.default);
    }

    return createElement(this.components.get("form"), {
      ref: this.ref,
      key: this.key,
      attrs: {
        id: this.id,
        name: this.name,
        role: this.search ? "search" : undefined
      },
      props: props,
      on: this.listeners
    }, nodes);
  },
  methods: {
    /**
     * Get the HTML form object reference.
     *
     * **Example**
     *
     * ```html
     * <template>
     *   <FormSchema ref="formSchema" :schema="schema"/>
     * </template>
     *
     * <script>
     *   import FormSchema from '@formschema/native'
     *
     *   export default {
     *     components: { FormSchema },
     *     data: () => ({
     *       schema: { type: 'string' }
     *     }),
     *     mounted() {
     *       console.log(this.$refs.formSchema.form())
     *     }
     *   };
     * </script>
     * ```
     *
     * @returns {HTMLFormElement|VNode|undefined} - An `HTMLFormElement` object or a `VNode` object describing the form element object, or `undefined` for input JSON schema object.
     */
    form(): HTMLFormElement | VNode | undefined {
      return this.$refs[this.ref] as any;
    },

    /**
     * @private
     */
    emitInputEvent(value: unknown, field: UnknowField): void {
      if (this.ready && this.validator) {
        this.validator(field).then((result) => {
          if (result) {
            this.ready = true;
          }
        });
      }

      /**
       * Fired synchronously when the value of an element is changed.
       */
      this.$emit("input", value);

      this.$nextTick(() => {
        this.ready = true;
      });
    },

    /**
     * @private
     */
    update([ field ]: UnknowField[]): void {
      this.key = field.key;
    }
  }
};

export default FormSchema;
