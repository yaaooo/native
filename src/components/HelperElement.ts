import { VNode } from "vue";
import { HelperComponent } from "../../types";

export const HelperElement: HelperComponent = {
  name: "HelperElement",
  functional: true,
  render(h, { props }): VNode | VNode[] {
    const field = props.field;
    const descriptor = props.field.descriptor;

    if (descriptor.helper) {
      const tag = field.isRoot ? "p" : "span";
      const data = {
        attrs: {
          ...descriptor.helperAttrs,
          "data-fs-helper": true
        }
      };

      return h(tag, data, descriptor.helper);
    }

    return null as any; // render nothing
  }
};
