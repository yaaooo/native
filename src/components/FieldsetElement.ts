import { VNode } from "vue";
import { FieldElement } from "@/components/FieldElement";
import { Objects } from "@/lib/Objects";
import { Fieldset } from "@/lib/Fieldset";
import { Field } from "@/lib/Field";
import { FieldsetComponent } from "../../types";

export const FieldsetElement: FieldsetComponent = {
  name: "FieldsetElement",
  functional: true,
  render(h, { data, props, slots }): VNode | VNode[] {
    const field = props.field;
    const descriptor = props.field.descriptor;

    const nodes = descriptor.kind === "object" && !Objects.isEmpty(descriptor.groups)
      ? Fieldset.renderGroups(h, field)
      : slots().default || Fieldset.renderChildren(h, field);

    if (descriptor.definition.kind === "hidden") {
      return nodes;
    }

    if (!field.isRoot) {
      return h(descriptor.definition.layout || FieldElement, data, nodes);
    }

    const attrs = { ...descriptor.attrs, ...(data.attrs || {}) };
    const helper = h(descriptor.components.get("helper"), data);

    // avoid rendering of <!-- --> by only confidering defined herper
    if (helper.tag) {
      nodes.unshift(helper);
    }

    if (descriptor.label && descriptor.layout === "fieldset") {
      const attrs = descriptor.labelAttrs;
      const legend = h("legend", { attrs }, descriptor.label);

      nodes.unshift(legend);
    }

    Field.renderMessages(h, field, nodes, field.isRoot);

    return h(descriptor.layout, { props, attrs }, nodes);
  }
};
