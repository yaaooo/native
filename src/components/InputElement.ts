import { VNode } from "vue";
import { CreateInput } from "@/lib/CreateInput";
import { FieldElement } from "@/components/FieldElement";
import { InputComponent, InputField } from "../../types";

export const InputElement: InputComponent = {
  name: "InputElement",
  functional: true,
  render(h, { data, children }): VNode | VNode[] {
    return h(FieldElement, data, [
      CreateInput<InputField>(h, "input", data),
      ...(children || [])
    ]);
  }
};
