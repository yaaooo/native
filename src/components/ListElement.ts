import { VNode } from "vue";
import { CreateInput } from "@/lib/CreateInput";
import { FieldElement } from "@/components/FieldElement";
import { ListComponent } from "../../types";

export const ListElement: ListComponent = {
  name: "ListElement",
  functional: true,
  render(h, { data, props }): VNode | VNode[] {
    const descriptor = props.field.descriptor;
    const children = descriptor.options.map(({ label, value, selected }) => {
      return h("option", { attrs: { value, selected } }, label);
    });

    return h(FieldElement, data, [
      CreateInput(h, "select", data, children, "change")
    ]);
  }
};
