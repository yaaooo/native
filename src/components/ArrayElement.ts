import { VNode, VNodeData } from "vue";
import { FieldElement } from "@/components/FieldElement";
import { FieldsetElement } from "@/components/FieldsetElement";
import { ArrayComponent } from "../../types";

export const ArrayElement: ArrayComponent = {
  name: "ArrayElement",
  functional: true,
  render(h, { data, props }): VNode | VNode[] {
    const ArrayButtonElement = props.field.descriptor.components.get("button");
    const nodes = props.field.children.map((childField) => {
      const childDescriptor = childField.descriptor;
      const buttonsWrapper: VNode[] = [];
      const childData: VNodeData = {
        key: childField.key,
        attrs: childDescriptor.attrs,
        props: { field: childField }
      };

      if (!props.field.uniqueItems && props.field.sortable) {
        if (childDescriptor.buttons.length && (props.field.schema.minItems === 0 || props.field.value.length > 1)) {
          const buttons = childDescriptor.buttons;
          const buttonsNodes = buttons.map((button) => h(button.component || ArrayButtonElement, {
            props: { button, field: childField }
          }));

          buttonsWrapper.push(h("div", {
            key: props.field.key + childField.key,
            attrs: {
              "data-fs-buttons": buttonsNodes.length
            }
          }, buttonsNodes));
        }
      }

      if (childDescriptor.kind === "object") {
        const componentNode = h(FieldsetElement, childData);
        const fieldsetData = {
          props: {
            // TODO this sounds deprecated. The field object no longer has a helper property
            field: { ...childField, helper: {} } // remove the duplicate bottom helper
          }
        };

        return h(FieldElement, fieldsetData, [ componentNode, buttonsWrapper ]);
      }

      return h(childDescriptor.component, childData, buttonsWrapper);
    });

    if (!props.field.uniqueItems && props.field.children.length < props.field.maxItems) {
      if (props.field.pushButton) {
        if (props.field.descriptor.definition.pushButton?.visible !== false) {
          nodes.push(h(props.field.descriptor.pushButton.component || ArrayButtonElement, {
            props: {
              button: props.field.descriptor.pushButton,
              field: props.field
            }
          }));
        }
      }
    }

    return h(FieldsetElement, data, nodes);
  }
};
