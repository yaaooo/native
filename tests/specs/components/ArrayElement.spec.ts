import { mount } from "@vue/test-utils";
import { ArrayElement } from "@/components/ArrayElement";
import { Options } from "../../lib/Options";

describe("components/ArrayElement", () => {
  it("should successfully render component with an empty array", () => {
    const { context } = Options.get({
      id: "id-characters",
      name: "characters",
      schema: {
        type: "array",
        title: "Empty array",
        description: "Your characters"
      }
    });

    const wrapper = mount(ArrayElement, { context });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it("should successfully render component with an empty array with a defined model", () => {
    const { context } = Options.get({
      id: "id-characters",
      name: "characters",
      schema: {
        type: "array",
        title: "Empty array",
        description: "Your characters"
      },
      model: [ "freezer" ]
    });

    const wrapper = mount(ArrayElement, { context });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it("should successfully render component with an array schema with an empty model", () => {
    const { context } = Options.get({
      id: "id-characters",
      name: "characters",
      schema: {
        type: "array",
        title: "Characters",
        description: "Your characters",
        minItems: 1,
        maxItems: 3,
        items: { type: "string" }
      },
      model: []
    });

    const wrapper = mount(ArrayElement, { context });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it("should successfully render component with an array schema with an empty model and minItem === maxItem", () => {
    const { context } = Options.get({
      id: "id-characters",
      name: "characters",
      schema: {
        type: "array",
        title: "Characters",
        description: "Your characters",
        minItems: 1,
        maxItems: 1,
        items: { type: "string" }
      },
      model: []
    });

    const wrapper = mount(ArrayElement, { context });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it("should successfully render component with an array schema with a model and value.length === 1 and minItem === 0", () => {
    const { context } = Options.get({
      id: "id-characters",
      name: "characters",
      schema: {
        type: "array",
        minItems: 0,
        items: { type: "string" }
      },
      model: [ "goku" ]
    });

    const wrapper = mount(ArrayElement, { context });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it("should successfully render component with an array schema with a model and value.length === 1", () => {
    const { context } = Options.get({
      id: "id-characters",
      name: "characters",
      schema: {
        type: "array",
        items: { type: "string" }
      },
      model: [ "goku" ]
    });

    const wrapper = mount(ArrayElement, { context });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it("should successfully render component with an array schema with a model and value.length > 1", () => {
    const { context } = Options.get({
      id: "id-characters",
      name: "characters",
      schema: {
        type: "array",
        items: { type: "string" }
      },
      model: [ "goku", "goten" ]
    });

    const wrapper = mount(ArrayElement, { context });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it("should successfully render component with an array schema with a defined model", () => {
    const { context } = Options.get({
      id: "id-characters",
      name: "characters",
      schema: {
        type: "array",
        title: "Characters",
        description: "Your characters",
        minItems: 1,
        maxItems: 3,
        items: { type: "string" }
      },
      model: [ "Goku", "Freezer" ]
    });

    const wrapper = mount(ArrayElement, { context });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it("should successfully render component without the push button when maxItems <= 1", () => {
    const { context } = Options.get({
      id: "id-characters",
      name: "characters",
      schema: {
        type: "array",
        title: "Characters",
        description: "Your characters",
        maxItems: 1,
        items: {
          type: "string"
        },
        default: [ "Goku" ]
      }
    });

    const wrapper = mount(ArrayElement, { context });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it("should successfully render component without the push button when maxItems <= 1 and minItems === 1", () => {
    const { context } = Options.get({
      id: "id-characters",
      name: "characters",
      schema: {
        type: "array",
        title: "Characters",
        description: "Your characters",
        minItems: 1,
        maxItems: 1,
        items: {
          type: "string"
        },
        default: [ "Goku" ]
      }
    });

    const wrapper = mount(ArrayElement, { context });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it("clicking to the push button should successfully update the parser model", () => {
    const { context, parser } = Options.get({
      id: "id-characters",
      name: "characters",
      schema: {
        type: "array",
        title: "Characters",
        description: "Your characters",
        minItems: 0,
        maxItems: 3,
        items: [
          { type: "string", default: "Goku" },
          { type: "string", default: "Freezer" }
        ]
      },
      model: []
    });

    const wrapper = mount(ArrayElement, { context });
    const button = wrapper.find("button[data-fs-button=push]");

    expect(parser.model).toEqual([]);

    button.trigger("click");
    expect(parser.model).toEqual([ "Goku" ]);

    button.trigger("click");
    expect(parser.model).toEqual([ "Goku", "Freezer" ]);
  });

  it("clicking to the push button should successfully trigger the options.onChange callback", () => {
    const { context, parser } = Options.get({
      id: "id-characters",
      name: "characters",
      schema: {
        type: "array",
        title: "Characters",
        description: "Your characters",
        minItems: 1,
        maxItems: 3,
        items: { type: "string" }
      },
      model: [ "Goku", "Freezer" ]
    });

    const wrapper = mount(ArrayElement, { context });
    const button = wrapper.find("button[data-fs-button=push]");

    expect(parser.model).toEqual([ "Goku", "Freezer" ]);
    expect(parser.rawValue).toEqual([ "Goku", "Freezer" ]);

    button.trigger("click");

    expect(parser.model).toEqual([ "Goku", "Freezer" ]);
    expect(parser.rawValue).toEqual([ "Goku", "Freezer", undefined ]);
  });

  it("clicking to the push button should successfully render a new array item", () => {
    const { context, parser } = Options.get({
      id: "id-characters",
      name: "characters",
      schema: {
        type: "array",
        title: "Characters",
        description: "Your characters",
        items: [
          { type: "string", default: "Goku", title: "First" },
          { type: "string", default: "Freezer", title: "Last" }
        ]
      },
      model: []
    });

    const wrapper = mount(ArrayElement, { context });
    const button = wrapper.find("button[data-fs-button=push]");

    // initial state: the fieldset is rendered with the push button
    expect(parser.model).toEqual([]);
    expect(wrapper.html()).toMatchSnapshot();

    // first push button click: render the first input
    button.trigger("click");
    expect(parser.model).toEqual([ "Goku" ]);
    expect(wrapper.html()).toMatchSnapshot();

    // second push button click: render the last input and disable the push button (no more item to add)
    button.trigger("click");
    expect(parser.model).toEqual([ "Goku", "Freezer" ]);
    expect(wrapper.html()).toMatchSnapshot();
  });

  describe("should successfully render enum array schema", () => {
    const onChangeMock = jest.fn();
    const { context, schema } = Options.get({
      id: "id-characters",
      name: "characters",
      schema: {
        type: "array",
        title: "Characters",
        description: "Your favorite characters",
        items: {
          type: "string",
          enum: [ "Goku", "Gohan", "Vegeta", "Picolo" ]
        },
        uniqueItems: true
      },
      model: [ "Goku", "Picolo" ],
      onChange: onChangeMock
    });

    const wrapper = mount(ArrayElement, { context });
    const pushButton = wrapper.find("button[data-fs-button=push]");
    const inputs: any = wrapper.findAll("input[type=checkbox]");

    it("should don't have push button", () => {
      expect(pushButton.exists()).toBeFalsy();
    });

    it("should have exaltly schema.items.enum.length input[type=checkbox]", () => {
      expect(inputs.length).toBe(schema.items.enum.length);
    });

    it("should have right initial model", () => {
      const [ [ initialValue ] ] = onChangeMock.mock.calls;

      expect(initialValue).toEqual([ "Goku", "Picolo" ]);
      expect(onChangeMock.mock.calls.length).toBe(1);
    });

    it("should have updated model after interaction", () => {
      inputs.at(1).setChecked(true);

      // eslint-disable-next-line no-empty-pattern
      const [ [], [ changedValue ] ] = onChangeMock.mock.calls;

      expect(changedValue).toEqual([ "Goku", "Gohan", "Picolo" ]);
      expect(onChangeMock.mock.calls.length).toBe(2);
    });
  });
});
