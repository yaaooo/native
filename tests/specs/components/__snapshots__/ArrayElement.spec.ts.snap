// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/ArrayElement clicking to the push button should successfully render a new array item 1`] = `
<fieldset id="id-characters" name="characters" aria-labelledby="id-characters-label" aria-describedby="id-characters-helper">
  <legend id="id-characters-label" for="id-characters">Characters</legend>
  <p id="id-characters-helper" data-fs-helper="true">Your characters</p><button type="button" data-fs-button="push">+</button>
</fieldset>
`;

exports[`components/ArrayElement clicking to the push button should successfully render a new array item 2`] = `
<fieldset id="id-characters" name="characters" aria-labelledby="id-characters-label" aria-describedby="id-characters-helper">
  <legend id="id-characters-label" for="id-characters">Characters</legend>
  <p id="id-characters-helper" data-fs-helper="true">Your characters</p><button type="button" data-fs-button="push">+</button>
</fieldset>
`;

exports[`components/ArrayElement clicking to the push button should successfully render a new array item 3`] = `
<fieldset id="id-characters" name="characters" aria-labelledby="id-characters-label" aria-describedby="id-characters-helper">
  <legend id="id-characters-label" for="id-characters">Characters</legend>
  <p id="id-characters-helper" data-fs-helper="true">Your characters</p><button type="button" data-fs-button="push">+</button>
</fieldset>
`;

exports[`components/ArrayElement should successfully render component with an array schema with a defined model 1`] = `
<fieldset id="id-characters" name="characters" aria-labelledby="id-characters-label" aria-describedby="id-characters-helper">
  <legend id="id-characters-label" for="id-characters">Characters</legend>
  <p id="id-characters-helper" data-fs-helper="true">Your characters</p>
  <div data-fs-kind="string" data-fs-type="text" data-fs-field="characters"><label for="id-characters-0"></label>
    <div data-fs-input="text"><input id="id-characters-0" type="text" name="characters" value="Goku">
      <div data-fs-buttons="3"><button type="button" disabled="disabled" data-fs-button="moveUp">↑</button><button type="button" data-fs-button="moveDown">↓</button><button type="button" data-fs-button="delete">-</button></div>
    </div>
  </div>
  <div data-fs-kind="string" data-fs-type="text" data-fs-field="characters"><label for="id-characters-1"></label>
    <div data-fs-input="text"><input id="id-characters-1" type="text" name="characters" value="Freezer">
      <div data-fs-buttons="3"><button type="button" data-fs-button="moveUp">↑</button><button type="button" disabled="disabled" data-fs-button="moveDown">↓</button><button type="button" data-fs-button="delete">-</button></div>
    </div>
  </div><button type="button" data-fs-button="push">+</button>
</fieldset>
`;

exports[`components/ArrayElement should successfully render component with an array schema with a model and value.length === 1 1`] = `
<fieldset id="id-characters" name="characters">
  <div data-fs-kind="string" data-fs-type="text" data-fs-field="characters"><label for="id-characters-0"></label>
    <div data-fs-input="text"><input id="id-characters-0" type="text" name="characters" value="goku">
      <div data-fs-buttons="3"><button type="button" disabled="disabled" data-fs-button="moveUp">↑</button><button type="button" disabled="disabled" data-fs-button="moveDown">↓</button><button type="button" data-fs-button="delete">-</button></div>
    </div>
  </div><button type="button" data-fs-button="push">+</button>
</fieldset>
`;

exports[`components/ArrayElement should successfully render component with an array schema with a model and value.length === 1 and minItem === 0 1`] = `
<fieldset id="id-characters" name="characters">
  <div data-fs-kind="string" data-fs-type="text" data-fs-field="characters"><label for="id-characters-0"></label>
    <div data-fs-input="text"><input id="id-characters-0" type="text" name="characters" value="goku">
      <div data-fs-buttons="3"><button type="button" disabled="disabled" data-fs-button="moveUp">↑</button><button type="button" disabled="disabled" data-fs-button="moveDown">↓</button><button type="button" data-fs-button="delete">-</button></div>
    </div>
  </div><button type="button" data-fs-button="push">+</button>
</fieldset>
`;

exports[`components/ArrayElement should successfully render component with an array schema with a model and value.length > 1 1`] = `
<fieldset id="id-characters" name="characters">
  <div data-fs-kind="string" data-fs-type="text" data-fs-field="characters"><label for="id-characters-0"></label>
    <div data-fs-input="text"><input id="id-characters-0" type="text" name="characters" value="goku">
      <div data-fs-buttons="3"><button type="button" disabled="disabled" data-fs-button="moveUp">↑</button><button type="button" data-fs-button="moveDown">↓</button><button type="button" data-fs-button="delete">-</button></div>
    </div>
  </div>
  <div data-fs-kind="string" data-fs-type="text" data-fs-field="characters"><label for="id-characters-1"></label>
    <div data-fs-input="text"><input id="id-characters-1" type="text" name="characters" value="goten">
      <div data-fs-buttons="3"><button type="button" data-fs-button="moveUp">↑</button><button type="button" disabled="disabled" data-fs-button="moveDown">↓</button><button type="button" data-fs-button="delete">-</button></div>
    </div>
  </div><button type="button" data-fs-button="push">+</button>
</fieldset>
`;

exports[`components/ArrayElement should successfully render component with an array schema with an empty model 1`] = `
<fieldset id="id-characters" name="characters" aria-labelledby="id-characters-label" aria-describedby="id-characters-helper">
  <legend id="id-characters-label" for="id-characters">Characters</legend>
  <p id="id-characters-helper" data-fs-helper="true">Your characters</p>
  <div data-fs-kind="string" data-fs-type="text" data-fs-field="characters"><label for="id-characters-0"></label>
    <div data-fs-input="text"><input id="id-characters-0" type="text" name="characters"></div>
  </div><button type="button" data-fs-button="push">+</button>
</fieldset>
`;

exports[`components/ArrayElement should successfully render component with an array schema with an empty model and minItem === maxItem 1`] = `
<fieldset id="id-characters" name="characters" aria-labelledby="id-characters-label" aria-describedby="id-characters-helper">
  <legend id="id-characters-label" for="id-characters">Characters</legend>
  <p id="id-characters-helper" data-fs-helper="true">Your characters</p>
  <div data-fs-kind="string" data-fs-type="text" data-fs-field="characters"><label for="id-characters-0"></label>
    <div data-fs-input="text"><input id="id-characters-0" type="text" name="characters"></div>
  </div>
</fieldset>
`;

exports[`components/ArrayElement should successfully render component with an empty array 1`] = `
<fieldset id="id-characters" name="characters" aria-labelledby="id-characters-label" aria-describedby="id-characters-helper">
  <legend id="id-characters-label" for="id-characters">Empty array</legend>
  <p id="id-characters-helper" data-fs-helper="true">Your characters</p><button type="button" disabled="disabled" data-fs-button="push">+</button>
</fieldset>
`;

exports[`components/ArrayElement should successfully render component with an empty array with a defined model 1`] = `
<fieldset id="id-characters" name="characters" aria-labelledby="id-characters-label" aria-describedby="id-characters-helper">
  <legend id="id-characters-label" for="id-characters">Empty array</legend>
  <p id="id-characters-helper" data-fs-helper="true">Your characters</p><button type="button" disabled="disabled" data-fs-button="push">+</button>
</fieldset>
`;

exports[`components/ArrayElement should successfully render component without the push button when maxItems <= 1 1`] = `
<fieldset id="id-characters" name="characters" aria-labelledby="id-characters-label" aria-describedby="id-characters-helper">
  <legend id="id-characters-label" for="id-characters">Characters</legend>
  <p id="id-characters-helper" data-fs-helper="true">Your characters</p>
  <div data-fs-kind="string" data-fs-type="text" data-fs-field="characters"><label for="id-characters-0"></label>
    <div data-fs-input="text"><input id="id-characters-0" type="text" name="characters" value="Goku">
      <div data-fs-buttons="3"><button type="button" disabled="disabled" data-fs-button="moveUp">↑</button><button type="button" disabled="disabled" data-fs-button="moveDown">↓</button><button type="button" data-fs-button="delete">-</button></div>
    </div>
  </div>
</fieldset>
`;

exports[`components/ArrayElement should successfully render component without the push button when maxItems <= 1 and minItems === 1 1`] = `
<fieldset id="id-characters" name="characters" aria-labelledby="id-characters-label" aria-describedby="id-characters-helper">
  <legend id="id-characters-label" for="id-characters">Characters</legend>
  <p id="id-characters-helper" data-fs-helper="true">Your characters</p>
  <div data-fs-kind="string" data-fs-type="text" data-fs-field="characters"><label for="id-characters-0"></label>
    <div data-fs-input="text"><input id="id-characters-0" type="text" name="characters" value="Goku">
      <div data-fs-buttons="3"><button type="button" disabled="disabled" data-fs-button="moveUp">↑</button><button type="button" disabled="disabled" data-fs-button="moveDown">↓</button><button type="button" data-fs-button="delete">-</button></div>
    </div>
  </div>
</fieldset>
`;
